pkg/expansion/expand.go

```diff
@@ -0,0 +1,224 @@
+			if err != nil {
+				return nil, fmt.Errorf("error expanding generator: %s", err)
+			}
+			resultants = append(resultants, result)
+		}
+	}
+
+	return resultants, nil
+}
+
+func ExpandGenerator(generator *unstructured.Unstructured, template *mutationsunversioned.TemplateExpansion, mutators []types.Mutator) (*unstructured.Unstructured, error) {
+	srcPath := template.Spec.TemplateSource
+	if srcPath == "" {
+		return nil, fmt.Errorf("cannot expand generator for template with no source")
+	}
+	resultantGVK := template.Spec.GeneratedGVK
+	emptyGVK := schema.GroupVersionKind{}
+	if resultantGVK == emptyGVK {
+		return nil, fmt.Errorf("cannot expand generator for template with empty generatedGVK")
+	}
+
+	src, ok, err := unstructured.NestedMap(generator.Object, sourcePath(srcPath)...)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract source field from unstructured: %s", err)
+	}
+	if !ok {
+		return nil, fmt.Errorf("could not find source field %q in generator", srcPath)
+	}
+
+	return createResultantKind(resultantGVK, src, mutators)
+}
+
+func convertUnstructuredToTyped(u *unstructured.Unstructured, obj interface{}) error {
+	if u == nil {
+		return fmt.Errorf("cannot convert nil unstructured to type")
+	}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), obj)
+	return err
+}
+
+func convertTemplateExpansion(u *unstructured.Unstructured) (mutationsunversioned.TemplateExpansion, error) {
+	te := mutationsunversioned.TemplateExpansion{}
+	err := convertUnstructuredToTyped(u, &te)
+	return te, err
+}
+
+func convertAssign(u *unstructured.Unstructured) (mutationsunversioned.Assign, error) {
+	a := mutationsunversioned.Assign{}
+	err := convertUnstructuredToTyped(u, &a)
+	return a, err
+}
+
+func convertAssignMetadata(u *unstructured.Unstructured) (mutationsunversioned.AssignMetadata, error) {
+	am := mutationsunversioned.AssignMetadata{}
+	err := convertUnstructuredToTyped(u, &am)
+	return am, err
+}
+
+func convertModifySet(u *unstructured.Unstructured) (mutationsunversioned.ModifySet, error) {
+	ms := mutationsunversioned.ModifySet{}
+	err := convertUnstructuredToTyped(u, &ms)
+	return ms, err
+}
+
+// sortResources sorts a list of resources into mutators, generators, template expansions and return
+// them respectively.
+func sortResources(resources []*unstructured.Unstructured) ([]*unstructured.Unstructured, []*unstructured.Unstructured, []*unstructured.Unstructured) {
+	var generators []*unstructured.Unstructured
+	var mutators []*unstructured.Unstructured
+	var templates []*unstructured.Unstructured
+
+	for _, r := range resources {
+		k := r.GetKind()
+		_, isMutator := MutatorTypes[k]
+		switch {
+		case isMutator:
+			mutators = append(mutators, r)
+		case k == "TemplateExpansion":
+			templates = append(templates, r)
+		default:
+			generators = append(generators, r)
+		}
+	}
+
+	return generators, mutators, templates
+}
+
+func sortMutators(mutators []types.Mutator) {
```

Comment:
In general, the mutator management code can go away and we can use the mutation system in `pkg/mutation/system.go` to manage/apply mutators, which will avoid code duplication.